presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -535,14 +535,47 @@ private void addSerializedCommitOutputToOutput(ConnectorCommitHandle commitHandl
         if (!output.get().isPresent()) {
             return;
         }
-
         Output outputInfo = output.get().get();
         SchemaTableName table = new SchemaTableName(outputInfo.getSchema(), outputInfo.getTable());
         output.set(Optional.of(new Output(
                 outputInfo.getConnectorId(),
                 outputInfo.getSchema(),
                 outputInfo.getTable(),
-                commitHandle.getSerializedCommitOutput(table))));
+                commitHandle.getSerializedCommitOutputForWrite(table))));
+    }
+
+    private void addSerializedCommitOutputToInputs(List<?> commitHandles)
+    {
+        ImmutableSet.Builder<Input> builder = ImmutableSet.builder();
+
+        for (Input input : inputs.get()) {
+            builder.add(attachSerializedCommitOutput(input, commitHandles));
+        }
+
+        inputs.set(builder.build());
+    }
+
+    private Input attachSerializedCommitOutput(Input input, List<?> commitHandles)
+    {
+        SchemaTableName table = new SchemaTableName(input.getSchema(), input.getTable());
+        for (int i = 0; i < commitHandles.size(); i++) {
+            if (!(commitHandles.get(i) instanceof ConnectorCommitHandle)) {
+                continue;
+            }
```

Comment:
Can we add a comment why we would get something beyond ConnectorCommitHandle?

Reply:
From the current code, it should be ConnectorCommitHandle. If it is not, we should throw. Maybe throwing an exception is better in this case.